src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -26,114 +26,198 @@
-                    return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
-                }
-                pet.setOwner(matchingOwners.iterator().next());
+            String action = fields[4];
+            if (action.equalsIgnoreCase("add")) {
+                petsToAdd.add(pet);
+            } else if (action.equalsIgnoreCase("delete")) {
+                petsToDelete.add(pet);
+            } else {
+                return errorResponse("invalid action");
             }
+        }
+
+        for (Pet p : petsToAdd) {
+            clinicService.savePet(p);
+        }
+
+        for (Pet p : petsToDelete) {
+            deletePet(p);
+        }
+
+        return new ResponseEntity<>(
+            Stream.concat(
+                petsToAdd.stream(),
+                petsToDelete.stream()
+            ).collect(Collectors.toList()),
+            HttpStatus.OK
+        );
+    }
 
-            if (csv.charAt(i) == ';') {
-                i++;
-
-                field = "";
-                while (i < csv.length() && csv.charAt(i) != '\n') {
-                    field += csv.charAt(i++);
-                }
-
-                if (field.toLowerCase().equals("add")) {
-                    clinicService.savePet(pet);
-                } else {
-                    for (Pet q : pet.getOwner().getPets()) {
-                        if (q.getName().equals(pet.getName())) {
-                            if (q.getType().getId().equals(pet.getType().getId())) {
-                                if (pet.getBirthDate().equals(q.getBirthDate())) {
-                                    clinicService.deletePet(q);
-                                }
-                            }
-                        }
-                    }
-                }
+    private static class SettingResponse {
+        public boolean isSuccessful;
+        public String errorMessage;
 
-            } else {
-                clinicService.savePet(pet);
-            }
-            i++;
+        public SettingResponse() {
+            this.isSuccessful = true;
+            this.errorMessage = null;
+        }
+
+        public SettingResponse(String errorMessage) {
+            this.isSuccessful = false;
+            this.errorMessage = errorMessage;
+        }
+    }
+
+    
+    /** errorResponse: Creates an Http message with a bad request and a user defined message (headerValue).
+     * @param headerValue
+     * @return ResponseEntity<List<Pet>>
+     */
+    private ResponseEntity<List<Pet>> errorResponse(String headerValue) {
+        HttpHeaders headers = new HttpHeaders();
+        headers.add("errors", headerValue);
+        return new ResponseEntity<>(headers, HttpStatus.BAD_REQUEST);
+    }
+
+    
+    /** setDateOfBirth: Set's birthday to a pet, by parsing a string into a date.
+     * @param pet
+     * @param dateOfBirth
+     * @return SettingResponse
+     */
+    private SettingResponse setDateOfBirth(Pet pet, String dateOfBirth) {
```

Comment:
I like that you extracted and created this and the below functions, which fulfill one single purpose, are small and you also kept the indentations small :)